src/lib/Export.hs

```diff
@@ -96,25 +95,45 @@ prepareFunctionForExport' cc f = do
             cont $ Nest (ExportResult (b:>ety)) Empty
 
-    exportArgRecon :: (EnvReader m, EnvExtender m) => Nest ExportArg n l -> m n (Abs (Nest IBinder) (ListE Block) n)
-    exportArgRecon topArgs = go [] topArgs
+    exportArgRecon
+      :: (EnvReader m, EnvExtender m)
+      => NaryPiType n
+      -> Nest ExportArg n l
+      -> m n (Abs (Nest IBinder) (ListE Block) n)
+    exportArgRecon (NaryPiType (NonEmptyNest tb tbs) _ _) topArgs =
+      go [] (Abs (Nest tb tbs) UnitE) topArgs
       where
         go :: (EnvReader m, EnvExtender m)
-           => [Block n] -> Nest ExportArg n l -> m n (Abs (Nest IBinder) (ListE Block) n)
-        go argRecons = \case
-          Empty -> return $ Abs Empty $ ListE argRecons
+           => [Block n]
+           -> Abs (Nest PiBinder) UnitE n
+           -> Nest ExportArg n l
+           -> m n (Abs (Nest IBinder) (ListE Block) n)
+        go argRecons dexArgs = \case
+          Empty -> case dexArgs of
+            Abs Empty UnitE -> return $ Abs Empty $ ListE argRecons
+            _ -> error "zip error: dex args should correspond 1-to-1 with export args"
           Nest (ExportArg _ b) bs ->
-            refreshAbs (Abs b (EmptyAbs bs)) \(v':>ety) (Abs bs' UnitE) -> do
-              (ity, block) <- typeRecon (sink ety) $ Var $ binderName v'
-              Abs ibs' allRecons' <- go (sinkList argRecons ++ [sink block]) bs'
-              return $ Abs (Nest (IBinder v' ity) ibs') allRecons'
-
-        typeRecon :: EnvExtender m => ExportType n -> Atom n -> m n (IType, Block n)
-        typeRecon ety v = case ety of
-          ScalarType sbt ->
-            return (Scalar sbt, Block (BlockAnn (BaseTy $ Scalar sbt) Pure) Empty v)
+            refreshAbs (Abs b (EmptyAbs bs)) \(v':>ety) (Abs bs' UnitE) ->
+              sinkM dexArgs >>= \case
+                Abs (Nest (PiBinder dexB dexArgTy _) restDexArgs) UnitE -> do
+                  -- get the current newtype-wrapped arg type
+                  -- pass it to type recon (who will use it to call `wrapWithNewtype` on the result
+                  -- to recur, wrap the var in the newtype and apply the subst
+                  (ity, block) <- typeRecon (sink ety) dexArgTy $ Var $ binderName v'
+                  dexArgAtom <- wrapNewtypes dexArgTy (Var $ binderName v')
+                  restDexArgs' <- applySubst (dexB@>SubstVal dexArgAtom) $ Abs restDexArgs UnitE
+                  Abs ibs' allRecons' <- go (sinkList argRecons ++ [sink block]) restDexArgs' bs'
+                  return $ Abs (Nest (IBinder v' ity) ibs') allRecons'
+                _ -> error "zip error: dex args should correspond 1-to-1 with export args"
+
+        typeRecon :: EnvExtender m => ExportType n -> Type n -> Atom n -> m n (IType, Block n)
+        typeRecon ety dexTy v = case ety of
+          ScalarType sbt -> do
+            resultWrapped <- wrapNewtypes dexTy v
+            return (Scalar sbt, Block (BlockAnn (BaseTy $ Scalar sbt) Pure) Empty resultWrapped)
```

Comment:
`resultWrapped` will not necessarily have a scalar base type if it was wrapped in a newtype. But you might be able to `AtomicBlock resultWrapped` your way out of it.

Reply:
Yes, this is a bug. Great catch.